specs/altair/sync-protocol.md

```diff
@@ -101,50 +104,83 @@ def get_subtree_index(generalized_index: GeneralizedIndex) -> uint64:
 ```
 
+### `get_active_header`
+
+```python
+def get_active_header(update: LightClientUpdate) -> BeaconBlockHeader:
+    # The "active header" is the header that the update is trying to convince us
+    # to accept. If a finalized header is present, it's the finalized header,
+    # otherwise it's the attested header
+    if update.finalized_header != BeaconBlockHeader():
+        return update.finalized_header
+    else:
+        return update.attested_header
+```
+
+### `get_safety_threshold`
+
+```python
+def get_safety_threshold(store: LightClientStore) -> uint64:
+    return max(
+        store.previous_max_active_participants,     
+        store.current_max_active_participants
+    ) // 2
+```
+
 ## Light client state updates
 
-A light client maintains its state in a `store` object of type `LightClientStore` and receives `update` objects of type `LightClientUpdate`. Every `update` triggers `process_light_client_update(store, update, current_slot)` where `current_slot` is the current slot based on some local clock.
+A light client maintains its state in a `store` object of type `LightClientStore` and receives `update` objects of type `LightClientUpdate`. Every `update` triggers `process_light_client_update(store, update, current_slot)` where `current_slot` is the current slot based on some local clock. `process_slot_for_light_client_store` is processed every time the current slot increments.
+
+#### `process_slot_for_light_client_store`
+
+```python
+def process_slot_for_light_client_store(store: LightClientStore, current_slot: Slot) -> None:
+    if current_slot % SAFETY_THRESHOLD_PERIOD == 0:
+        store.previous_max_active_participants = store.current_max_active_participants
+        store.current_max_active_participants = 0
+```
 
 #### `validate_light_client_update`
 
 ```python
-def validate_light_client_update(snapshot: LightClientSnapshot,
+def validate_light_client_update(store: LightClientStore,
                                  update: LightClientUpdate,
+                                 current_slot: Slot,
                                  genesis_validators_root: Root) -> None:
-    # Verify update slot is larger than snapshot slot
-    assert update.header.slot > snapshot.header.slot
+                                 
+    # Verify update slot is larger than slot of current best finalized header
+    active_header = get_active_header(update)
+    assert current_slot >= active_header.slot > store.finalized_header.slot
 
     # Verify update does not skip a sync committee period
-    snapshot_period = compute_epoch_at_slot(snapshot.header.slot) // EPOCHS_PER_SYNC_COMMITTEE_PERIOD
-    update_period = compute_epoch_at_slot(update.header.slot) // EPOCHS_PER_SYNC_COMMITTEE_PERIOD
-    assert update_period in (snapshot_period, snapshot_period + 1)
+    finalized_period = compute_epoch_at_slot(store.finalized_header.slot) // EPOCHS_PER_SYNC_COMMITTEE_PERIOD
```

Comment:
Does that mean the light-client should fail if there is a skip period? This seems to be a fairly normal path when a client stops running for a few das.

Reply:
The light-client can request historic `LightClientUpdate` from the network. It needs at least one update per period to follow along, as it only knows `current_sync_committee` and `next_sync_committee` and can only verify `LightClientUpdate` from those periods.

However, what is still suboptimal is the case where `finalized_update` is in a different period than `attested_update`, but this is not a problem introduced by this PR. Another tricky case to tackle for the case where `attested_update` is in a different period than the committee which signed it, which probably even requires some heuristics to figure out (as this case depends on there being missed slots at the start of an epoch). For now, these edge cases are all ignored, and the updates are only accepted if all of `finalized_update`, `attested_update`, and the sync committee signing it come from the same sync committee period.

Reply:
Yes. agree on the updating path will trace sync-comm linkage. Also agree that this is not the issue raised by this PR.

With regard to the edge case...it could cause  some weird behaviors temporarily. For example, `apply_light_client_update` is called due to timeout. Then, when there are valid `update.finalized_header` arrives, they will get rejected.

Again, this behavior could be better handled if we assume that the light-client can make request for specific LightClientUpdates when it times out or fall out of sync with the current stream of updates. The sync logic would becomes a lot cleaner to be separated into two sync mode: skip-sync mode and normal sync mode.